packages/core/src/middleware/koa-quota-guard.ts

```diff
@@ -0,0 +1,74 @@
+import type { MiddlewareType } from 'koa';
+
+import { EnvSet } from '#src/env-set/index.js';
+import RequestError from '#src/errors/RequestError/index.js';
+import { type CloudConnectionLibrary } from '#src/libraries/cloud-connection.js';
+import type Queries from '#src/tenants/Queries.js';
+import assertThat from '#src/utils/assert-that.js';
+import { getTenantSubscriptionPlan } from '#src/utils/subscription/index.js';
+import { type SubscriptionQuota } from '#src/utils/subscription/types.js';
+
+type UsageGuardConfig = {
+  key: keyof SubscriptionQuota;
+  cloudConnection: CloudConnectionLibrary;
+  queries: Queries;
+};
+
+const getTenantUsage = async (key: keyof SubscriptionQuota, queries: Queries): Promise<number> => {
+  if (key === 'applicationsLimit') {
+    return queries.applications.countNonM2MApplications();
+  }
+
+  // TODO: add other keys
+
+  throw new Error('Unsupported subscription quota key');
+};
+
+export default function koaQuotaGuard<StateT, ContextT, ResponseBodyT>({
+  key,
+  queries,
+  cloudConnection,
+}: UsageGuardConfig): MiddlewareType<StateT, ContextT, ResponseBodyT> {
+  return async (ctx, next) => {
+    const { isCloud, isIntegrationTest, isProduction } = EnvSet.values;
+
+    // Disable in production until price feature is ready
+    if (!isCloud || isIntegrationTest || isProduction) {
+      return next();
+    }
+
+    const plan = await getTenantSubscriptionPlan(cloudConnection);
+    const limit = plan.quota[key];
+
+    if (typeof limit === 'boolean') {
+      assertThat(
+        limit,
```

Comment:
should also check the current value? seems the quota guard should not only rely on the quota limit?

Reply:
same as above, this is the guard that preventing from "adding" one